File: coauthor-python/coauthor/housekeeping/pack.py
import os
import shutil

from ..logger import logger
from ..utils.file import deleteFile, moveFile, findFile

from .constants import PACK_EXTENSIONS, TEMP_EXTENSIONS, HISTORY_DIR
from .utils import getAgent_first_name_chunk, get_file_patterns, get_folder_datetime


def runPackSingle(model: str, inputFile: str, agent: str, outputFolder: str | None = None) -> None:
    """Pack LaTeX files and related outputs into timestamped history directory, cleaning temp files."""
    baseName = os.path.splitext(os.path.basename(inputFile))[0]
    inputDir = os.path.dirname(inputFile)

    agent_first_name_chunk = getAgent_first_name_chunk(agent)

    file_patterns = get_file_patterns(baseName, model, agent_first_name_chunk)
    file_patterns.append(baseName)

    moved_files = []
    copied_files = []
    for pattern in file_patterns:
        for ext in PACK_EXTENSIONS:
            filePath = findFile(inputDir, pattern, ext)
            if filePath:
                if filePath == inputFile or pattern == baseName:
                    copied_files.append(filePath)
                else:
                    moved_files.append(filePath)

    # this includes the original input file, f"{base}_{agent}_r{round}_{model}",
    # so even if no output file from llm is genereated, the output folder will still be created
    if moved_files or copied_files:
        now = get_folder_datetime(inputDir, file_patterns, PACK_EXTENSIONS)
        if outputFolder is None:
            outputFolder = os.path.join(inputDir, HISTORY_DIR, f"{now}_{baseName}_{agent}_{model}")
        os.makedirs(outputFolder, exist_ok=True)
        for filePath in moved_files:
            moveFile(filePath, outputFolder)
        for filePath in copied_files:
            shutil.copy(filePath, outputFolder)
            logger.info(f"Copied file: {filePath}")

        logger.info(f"Files packed to: {outputFolder}")

    for pattern in file_patterns:
        for ext in TEMP_EXTENSIONS:
            filePath = findFile(inputDir, pattern, ext)
            if filePath and filePath != inputFile:
                deleteFile(filePath)

    logger.info(f"Packing finished: {inputFile}.")
    return outputFolder


def runPackMultiple(model: str, inputFile: str, inputFiles: list[str], agent: str, outputNameOverride: str | None = None) -> None:
    """Pack multiple LaTeX files and their outputs into a single history directory."""
    # Initialize baseName and output_dir
    if outputNameOverride:
        baseName = os.path.splitext(os.path.basename(outputNameOverride))[0]
        output_dir = os.path.dirname(outputNameOverride)
    elif inputFile:
        baseName = os.path.splitext(os.path.basename(inputFile))[0]
        output_dir = os.path.dirname(inputFile)
    else:
        raise ValueError("Either inputFile or outputNameOverride must be provided")

    agent_first_name_chunk = getAgent_first_name_chunk(agent)
    file_patterns = get_file_patterns(baseName, model, agent_first_name_chunk)

    # Add patterns for additional XML files
    additional_patterns = [f"{baseName}_{agent_first_name_chunk}_r0_{model}.xml", f"{baseName}_{agent_first_name_chunk}_r1_{model}.xml"]
    file_patterns.extend(additional_patterns)

    now = get_folder_datetime(output_dir, file_patterns, PACK_EXTENSIONS)
    commonOutputFolder = os.path.join(output_dir, HISTORY_DIR, f"{now}_{baseName}_multiple_{agent}_{model}")

    # Ensure the output folder exists
    os.makedirs(commonOutputFolder, exist_ok=True)

    # Pack input files
    for inputFile in inputFiles:
        logger.info(f"\nPacking files to: {commonOutputFolder}")
        runPackSingle(model, inputFile, agent, outputFolder=commonOutputFolder)

    # Pack additional XML files
    for pattern in additional_patterns:
        filePath = os.path.join(output_dir, pattern)
        if os.path.exists(filePath):
            moveFile(filePath, commonOutputFolder)

    logger.info(f"All files packed to: {commonOutputFolder}")
    return commonOutputFolder

File: src/housekeeping/pack.ts
// Standard library imports
import * as path from 'node:path';

// Third-party imports
import { Effect } from 'effect';

// Internal imports
import { withLogChannel, withLogData } from '@logger/effectLog';
import { WorkspaceFs } from '@platform/rootedFs';
import { getCleanAgentName, type FileOpResult } from '@shared/schemas';
import { toErrorMessage } from '@utils/errors/errorMessage';
import { copyFileExclusive } from '@utils/files/fsDurability';

// Local file imports
import {
  PACK_EXTENSIONS,
  TEMP_EXTENSIONS,
  HISTORY_DIR,
  CHANNEL,
} from './constants';
import { generateTimestamp, collectFilesFromPatterns } from './utils';

/** Every pack failure reaches the host as the same result shape. */
const asErrorResult = (error: unknown) =>
  Effect.logError('Error during file operations').pipe(
    withLogData(error),
    withLogChannel(CHANNEL),
    Effect.as<FileOpResult>({ status: 'error', error: toErrorMessage(error) }),
  );

/**
 * Pack the source document's own files (`<base>.<ext>`, e.g. its PDF) into a
 * `History/` folder by copy, then sweep the source basename's build
 * artifacts. Workflow outputs live in run storage and are packed by
 * `runPackRunDir`; no workspace filename is parsed for them here.
 */
export const runPackSingle = Effect.fn('housekeeping.runPackSingle')(function* (
  model: string,
  inputFile: string,
  agent: string,
  outputFolder?: string,
) {
  yield* Effect.logInfo(
    `Starting packing with model=${model}, inputFile=${inputFile}, agent=${agent}, outputFolder=${outputFolder}`,
  ).pipe(withLogChannel(CHANNEL));

  if (!inputFile || !model || !agent) {
    yield* Effect.logError(
      `Missing required parameters: model=${model}, inputFile=${inputFile}, agent=${agent}`,
    ).pipe(withLogChannel(CHANNEL));
    return { status: 'missingParams' } satisfies FileOpResult;
  }

  const workspaceFs = yield* WorkspaceFs;
  const workspaceRoot = workspaceFs.root;
  if (!workspaceRoot) return { status: 'noFiles' } satisfies FileOpResult;

  const baseName = path.parse(inputFile).name;
  const inputDir = path.dirname(inputFile);

  return yield* Effect.gen(function* () {
    const copiedFiles = [
      ...(yield* collectFilesFromPatterns(
        workspaceRoot,
        inputDir,
        [baseName],
        PACK_EXTENSIONS,
      )),
    ];

    // Nothing to pack when the only match (if any) is the input document.
    if (copiedFiles.every((file) => file === inputFile)) {
      yield* Effect.logWarning(`No files found to pack for ${inputFile}`).pipe(
        withLogChannel(CHANNEL),
      );
      return { status: 'noFiles' } satisfies FileOpResult;
    }

    yield* Effect.logDebug(`Files to copy:\n${copiedFiles.join('\n')}`).pipe(
      withLogChannel(CHANNEL),
    );

    const cleanAgent = getCleanAgentName(agent);
    const resolvedOutputFolder =
      outputFolder ||
      path.join(
        inputDir,
        HISTORY_DIR,
        `${generateTimestamp()}_${baseName}_${cleanAgent}_${model}`,
      );

    yield* workspaceFs.makeDirectory(resolvedOutputFolder, {
      recursive: true,
    });
    yield* Effect.logDebug(
      `Created output directory: ${resolvedOutputFolder}`,
    ).pipe(withLogChannel(CHANNEL));

    for (const file of copiedFiles) {
      const destination = path.join(resolvedOutputFolder, path.basename(file));
      yield* Effect.logDebug(`Copying: ${file} -> ${destination}`).pipe(
        withLogChannel(CHANNEL),
      );
      // Exclusive: the check and the creation are one step, so when
      // `runPackMultiple` packs two same-basename sources into one folder
      // concurrently, the losing copy fails with `AlreadyExists` and its pack
      // reports the collision instead of silently omitting its file.
      yield* copyFileExclusive(
        yield* workspaceFs.resolve(file),
        yield* workspaceFs.resolve(destination),
      );
    }
    yield* Effect.logInfo(`Files packed into ${resolvedOutputFolder}`).pipe(
      withLogChannel(CHANNEL),
    );

    // The source's own `<base>.bib` and `<base>.bak*` are the user's files,
    // not build artifacts: only sweep the source basename's generated ones.
    const packed = new Set(copiedFiles);
    const sweepable = yield* collectFilesFromPatterns(
      workspaceRoot,
      inputDir,
      [baseName],
      TEMP_EXTENSIONS.filter((ext) => ext !== '.bib' && ext !== '.bak*'),
    );
    for (const file of sweepable) {
      if (!packed.has(file)) {
        yield* workspaceFs.remove(file, { force: true });
      }
    }

    return {
      status: 'success',
      outputFolder: resolvedOutputFolder,
    } satisfies FileOpResult;
  }).pipe(Effect.catch(asErrorResult));
});

export const runPackMultiple = Effect.fn('housekeeping.runPackMultiple')(
  function* (
    model: string,
    inputFile: string,
    agent: string,
    inputFiles: string[],
  ) {
    yield* Effect.logDebug(
      `Starting multiple packing with model=${model}, inputFile=${inputFile}, agent=${agent}; additional files: ${inputFiles.join(', ')}`,
    ).pipe(withLogChannel(CHANNEL));

    const baseName = path.parse(inputFile).name;
    const outputDir = path.dirname(inputFile);
    const cleanAgent = getCleanAgentName(agent);
    const commonOutputFolder = path.join(
      outputDir,
      HISTORY_DIR,
      `${generateTimestamp()}_${baseName}_multiple_${cleanAgent}_${model}`,
    );

    const allFilesToPack = [inputFile, ...inputFiles];
    const results = yield* Effect.forEach(
      allFilesToPack,
      (file) => runPackSingle(model, file, agent, commonOutputFolder),
      { concurrency: 'unbounded' },
    );

    // A per-file error takes precedence over partial success, so a failed
    // pack is never misreported as success or as 'noFiles'.
    const errored = results.find((result) => result.status === 'error');
    if (errored) return errored;

    if (results.some((result) => result.status === 'success')) {
      yield* Effect.logInfo(`All files packed into ${commonOutputFolder}`).pipe(
        withLogChannel(CHANNEL),
      );
      return {
        status: 'success',
        outputFolder: commonOutputFolder,
      } satisfies FileOpResult;
    }

    yield* Effect.logWarning(`No files found to pack for ${inputFile}`).pipe(
      withLogChannel(CHANNEL),
    );
    return { status: 'noFiles' } satisfies FileOpResult;
  },
);
